### ISSUE TITLE
Successfully using space bar as a modifier question

### ISSUE THREAD

**Issue by divansantana** [alols/xcape#81]
More of a question. xcape is awesome, btw - thanks!

So `man xcape` says.
```
As an example, we can make the space bar work as an additional ctrl key when held with the following sequence of commands:
```
I've used `xcape` and `xmodmap` to make space key act as `mod4`/`Super_L`. It's easier to push space bar then the "windows" key. I use `i3` so this is pretty helpful too.

It's a nice change. But I find it really difficult to keep this change because now a space character is only inserted when the spacebar is fully released. When one types quickly I often land up with no spaces separated words and other similar issues.

I guess I'm wondering feedback on other users experience with mapping the spacebar as a modifier. Did you manage to it successfully and how?

**Comment by divansantana** [comment]
For me, I can't use modify the spacebar successfully and adapt to the new way it works without it interfering a lot. A better fix for me is to modify another "normal" key like `;` to work as `mod4`.

Only issue there is I haven't figured out how to do that with `xcape`. When I do this I can't get the `shift+;` to generate a colon.

**Comment by xiongtx** [comment]
I've had the same experience. 

The fundamental problem is that we are used to the space being inserted at key *press*, not *release*. Fast typist like us don't wait for the space bar to be released before hitting the next key. Trying to use space bar as a modifier key, however, *requires* that it is treated as Control until release. 

There's no way, even in theory, to get around this.

**Comment by divansantana** [comment]
@xiongtx Yes. For now I've mostly adapter to the issue. Using space for me has greatly helped with RSI. There's no way to resolve the issue. But is nice to share experiences etc.

Perhaps one could use another less used key like `,` instead of space, which is still easily in reach.

**Comment by diego-alvarez-hs** [comment]
:disappointed: I couldn't adapt to this workflow, I kept hitting other keys without space being released

**Comment by amosbird** [comment]
I use a japanese keyboard to overcome this.

**Comment by CNG** [comment]
I'm also working on adapting to this. It's been a bit frustrating when I was accidentally triggering various i3 commands during my typing, but it seems to be less frequent now. Perhaps I'm also "helped" by my typing being much slower lately [since switching to Dvorak](https://votecharlie.com/blog/2017/04/switching-to-dvorak.html).

I'm also experiencing the need to kill and restart xcape every time I plug or unplug a monitor; I am still tracking down this problem.

**Comment by CNG** [comment]
I'm now trying to stop using space as a modifier, as it's gotten too annoying to be accidentally launching programs and moving windows while typing and sharing my screen during meetings. Perhaps if there were a time threshold where xcape would change the spacebar to a modifier key only *after* it's been held down for some amount of time, that could solve this issue.

**Comment by divansantana** [comment]
My crazy setup. Which works well for me.

Caps is mod4/windows/super key. Easier to reach then the default mod4
place and used for my desktop (EXWM) shortcuts.

space bar is space and CTRL. I can't recall when last I hit the space
bar/CTRL key and didn't get what I expected. It took a few weeks to
adapt.

Mostly forced due to RSI issues.

```bash
setxkbmap -layout dvorak

pkill -9 -f xcape

# Make typematic keyboard delay and rate faster.
# default is 660ms and 25Hz
xset r rate 180 120

# defaults
# keycode 133 (keysym 0xffeb, Super_L) left window key
# keycode 65 (keysym 0x20, space) space
# keycode 66 (keysym 0xffe5, Caps_Lock)
# keycode 37 (keysym 0xffe3, Control_L)
# keycode 105 (keysym 0xffe4, Control_R)
# keycode 36 (keysym 0xff0d, Return)
# keycode 47 (keysym 0x3b, semicolon)
# keycode 9 (keysym 0xff1b, Escape)
# keycode 108 (keysym 0xffea, Alt_R)
# keycode 64 (keysym 0xffe9, Alt_L)

xmodmap -e "clear lock"
xmodmap -e "keycode 66 = Super_L"
xmodmap -e "keycode 36 = Super_R"

spare_modifier="Hyper_L"
xmodmap -e "keycode 65 = $spare_modifier"
xmodmap -e "remove mod4 = $spare_modifier"
xmodmap -e "add Control = $spare_modifier"

xmodmap -e "keycode any = space"
xmodmap -e "keycode any = Return"

xcape -t 1000 -e "Shift_L=Escape;Shift_R=Escape;Super_R=Return;$spare_modifier=space"

notify-send -t 5000 "Keyboard settings applied."
```

**Comment by forgetmenotz** [comment]
Haha, I was hoping I'd find other people with RSI and crazy remaps.

I have solved the space problem (at least satisfactorily enough for me). You'll notice that I have keys as triple functions (tap, hold, shifted). So for example, capslock is return, ISO_Level3_Shift, and backspace respectively. More about that later...

caps is return, backspace and Mode_switch
; is escape, and shift.
Shift_L is escape and control
\ is \ and Mode_switch 
space is space and ISO_Level3_Shift

What this means is that when I hold shift (using the ; key) space becomes ISO_Level3_Shift, which puts me in level6 (the sixth keysym after a given keycode in xmodmap, so in xmodmap "keycode 38 = a A a a a F26" when holding shift+space would make the a key = F26).

This method negates the problem of the typing mistakes, because the space only changes when shifted, which is not nearly as often; it does require holding two keys, in my case, ; and space together, but I find it to be a useful and comfortable alternative tot pressing alt. I just bind all the keys on the sixth level to F keys (F13 and above) then use them for binds in xbindkeys or sxhkd or in the wm config (for me i3).

Also, ISO_Level3_Shift changes all the alphabet keys to punctuation.
q=£ (we call pounds quids here in uk) w=* e== r=^ t=` y=? u=_ a=~ s=; d=$ f=% g=@ h=# z=0 x=! c=: v=& b=" n=+ m=- i=|  and I use the vim keys for the brackets/braces and o and p for open parenthesis.

Note that for the triple function capslock key to function (i.e. while holding shift) I had to use a forked version of xcape to disable it when shifted (can be found by searching for shift in the issues here) and added the ; key to be recognized as shift (an easy tinker in xcape.c)

I guess some people would think the punctuation mode on caps lock is a mess, but that's what I like to use. I never really learned to use the keyboard as its designed to be use I guess, and find this ergonomic.

**Comment by forgetmenotz** [comment]
@divansantana 

Interesting that you got used to it. Do you think you could use it if it was something other than control? I would be surprised if you didn't get capital letters if the space was xcaped to shift rather than control.

**Comment by danr** [comment]
Are you on traditional laptop keyboards? I have to use an "ergonomic" keyboard like the kinesis advantage. These give a lot of keys on the thumbs, and I use xcape on backspace and delete (which are thumb keys) for ctrl and alt. RSI kicks in on laptop keyboards after just a day of use or so.

**Comment by divansantana** [comment]
@forgetmenotz I'm getting about 70wpm with virtually no mistakes. I did learn to type from scratch again and that helped because I was typing slower and did have to adjust to the space thing.

I also noticed I only use my right thumb to press the space bar, which is not ideal but works.

**Comment by divansantana** [comment]
I mostly type on an ergodox, but I do type on a laptop dell keyboard often and havent't noticed any issues.

**Comment by forgetmenotz** [comment]
I had an idea that I thought might be able to solve this issue for those of us who hit the space modifier too often. If an xcape modifier could be set to disabled until a certain time after other keys are pressed. Have you ever used vim and lowered the time that the program waits for an escape sequence

**Comment by forgetmenotz** [comment]
I've been using this for a few weeks now. In favour of the increased functionality of having space as altgr, I've kind of accepted the slower typing speed, which unfortunately has not gone away, and I still make mistakes here and there.

**Comment by jirassimok** [comment]
> Trying to use space bar as a modifier key, however, requires that it is treated as Control until release. There's no way, even in theory, to get around this.

There is a way around this, but not with xcape, or any other tool that I am aware of for Linux. The solution is using different behaviors depending on whether the other keys are pressed _and_ released while holding the space bar.

The sequence `space down, X down, X up, space up`, should act like pressing `modifier`+`X`, while `space down, X down, space up, X up` should act like `space` `modifier`.

This mapping still reduces typing speed, but it is _much_ easier to learn than if both sequences turn `space` into a modifier, and much less prone to errors.

**Comment by MaxGyver83** [comment]
> There is a way around this, but not with xcape, or any other tool that I am aware of for Linux.

[KMonad](https://github.com/david-janssen/kmonad) can do that. Search the [tutorial](https://github.com/david-janssen/kmonad/blob/master/keymap/tutorial.kbd) for `tap-next-release`.
